triggers/tests/daml/Retry.daml

```diff
@@ -0,0 +1,65 @@
+-- Copyright (c) 2019 The DAML Authors. All rights reserved.
+-- SPDX-License-Identifier: Apache-2.0
+
+daml 1.2
+module Retry where
+
+import DA.Action
+import DA.Next.Map (Map)
+import qualified DA.Next.Map as Map
+import DA.Optional
+
+import Daml.Trigger
+
+retryTrigger = runTrigger Trigger
+  { initialize = \_acs -> 3
+  , updateState = \_acs msg allowedFail -> case msg of
+      MCompletion c
+        | Failed {} <- c.status -> allowedFail - 1
+      _ -> allowedFail
+  , rule = retryRule
+  }
+
+-- We first create a T template, then we try to exercise C 3 times until allowedRetries is 0
+-- finally we create Done
+retryRule : Party -> ACS -> Map CommandId [Command] -> Int -> TriggerA ()
+retryRule party acs commandsInFlight allowedRetries
+  | [] <- getTemplates @T acs = do
+    let create = T { p = party }
+    let pending = concatMap (mapOptional toCreate . snd) $ Map.toList commandsInFlight
+    when (create `notElem` pending) $ void $ emitCommands [createCmd create]
```

Comment:
Can this bit be factored out? It might even be useful as part of the triggers library. Checking if a command is already in flight seems like a common operation.

Reply:
Yeah, while writing this I also thought that this should be part of the high-level API. So something like `deduplicateCreate` and `deduplicateExercise`. I’ll leave that for a separate PR.